server/src/main/scala/org/scassandra/server/actors/BatchHandler.scala

```diff
@@ -1,42 +1,86 @@
-import org.scassandra.server.cqlmessages.{BatchType, Consistency, CqlProtocolHelper, CqlMessageFactory}
+import akka.util.{Timeout, ByteString}
+import akka.pattern.{ask, pipe}
+
+import org.scassandra.server.actors.BatchHandler.{BatchToFinish, Execute}
+import org.scassandra.server.actors.PrepareHandler.{PreparedStatementResponse, PreparedStatementQuery}
+import org.scassandra.server.cqlmessages._
 import org.scassandra.server.priming.{BatchExecution, BatchQuery, ActivityLog}
 
-class BatchHandler(tcpConnection: ActorRef, msgFactory: CqlMessageFactory, activityLog: ActivityLog) extends Actor with ActorLogging {
+import scala.collection.immutable.IndexedSeq
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+class BatchHandler(tcpConnection: ActorRef,
+                   msgFactory: CqlMessageFactory,
+                   activityLog: ActivityLog,
+                   prepareHandler: ActorRef) extends Actor with ActorLogging {
 
   import CqlProtocolHelper._
+  import context.dispatcher
+  implicit val timeout: Timeout = 1 second
 
-  override def receive: Receive = {
+  def receive: Receive = {
     case Execute(body, stream) =>
       val iterator = body.iterator
       val batchType = BatchType.fromCode(iterator.getByte)
       val numStatements = iterator.getShort
 
-      val statements: List[BatchQuery] = (0 until numStatements).map(_ => {
-        val kind = iterator.getByte
-        assert(kind == 0) // query rather than prepared statement
-        val query: String = readLongString(iterator).get
-        val numVariables = iterator.getShort
-        // read off the bytes for each variable, we can't parse them until priming of batches is supported
-        (0 until numVariables).foreach { _ =>
-          val throwAway = readLongBytes(iterator)
-          log.debug("throwing away bytes from batch statement variable as priming not supported yet {}", throwAway)
+      val statements: IndexedSeq[InFlightBatchQuery] = (0 until numStatements).map(_ => {
+        BatchQueryKind.fromCode(iterator.getByte) match {
+          case QueryKind =>
+            NormalQuery(msgFactory.parseBatchQuery(iterator))
+          case PreparedStatementKind =>
+            iterator.drop(2) // short representing the length of the id
+            val id = iterator.getInt
+            val variables = msgFactory.readVariables(iterator)
+            IncompletePreparedStatement(id, variables)
         }
-        BatchQuery(query)
-      }).toList
-
+      })
       val consistency = Consistency.fromCode(iterator.getShort)
 
-      activityLog.recordBatchExecution(BatchExecution(statements, consistency, batchType))
+      val ids: List[Int] = statements.foldLeft(List[Int]())((acc, next) => next match {
+        case IncompletePreparedStatement(id, _) => id :: acc
+        case _ => acc
+      })
+
+      log.debug("Prepared statement ids {}", ids)
+
+      if (ids.isEmpty) {
```

Comment:
Regarding UNPREPARED, definitely. I made nearly the same comment on your PR yesterday as I didn't realise it existed.

I think once the Execute is removed from the PrepareHandler it will be nicer.